src/gasPriceOracle/oracle.ts

```diff
@@ -14,35 +14,36 @@ import * as polygon from "./adapters/polygon";
  * @param legacyFallback In the case of an unrecognised chain, fall back to type 0 gas estimation.
  * @returns Am object of type GasPriceEstimate.
  */
-export async function getGasPriceEstimate(
-  provider: providers.Provider,
-  chainId?: number,
-  legacyFallback = true
-): Promise<GasPriceEstimate> {
-  if (chainId === undefined) {
-    ({ chainId } = await provider.getNetwork());
-  }
+export async function getGasPriceEstimate(provider: providers.Provider, transport?: Transport): Promise<GasPriceEstimate> {
+  const { chainId } = await provider.getNetwork();
+
+  transport ??= http(); // @todo: Inherit URL from provider.
+  const chain = extractChain({ chains: Object.values(chains), id: chainId as any });
```

Comment:
@gsteenkamp89 @dohaki Headache #1 - `extractChain()` apparently requires `id` as an explicit union type, rather than as a standard number. I bodged it here by casting to `any`, but that sucks.

Reply:
@pxrl You can opt out of viem's narrow type inference when some of your values are dynamic. Like here, your `chainId` is not strictly known at compile time so it is just a `number`. 

Don't use viem's `extractChain` utility. It is just `Array.find` with static param checking
```ts
const chain: Chain | undefined = Object.values(chains).find((chain) => chain.id === chainId);
```

But it is useful to note that the `PublicClient` type exported by viem doesn't have to be super narrow to access the necessary methods you would likely use. It doesn' have to know which chain you're on to fetch a txReceipt for example, it just has to know that _a_ chain has been configured in this client. 

So you can build this type:

```ts
export type ConfiguredPublicClient = PublicClient<Transport, Chain>;
```

So what I would do is this:

```ts
export type ConfiguredPublicClient = PublicClient<Transport, Chain>;

function getViemPublicClient(chainId: number, transport?: Transport): ConfiguredPublicClient {
  // don't use viem's extractChain utility when using dynamic chainIds
  const chain: Chain | undefined = Object.values(chains).find((chain) => chain.id === chainId);
  if (!chain) {
    throw new Error("Chain config not found");
  }
  // default to public http endpoint
  transport ??= http();
  return createPublicClient({ chain, transport });
}
```

Reply:
Another option would be to write a type guard that will narrow the type of the chainId such that it is in fact a chainId of one of the members of the `chains` union. This way you can get nicer type inference, but this is probably not needed, but here is how I would do it

```ts
function isSupportedChainId<Chains extends Chain[], Id extends Chains[number]["id"]>(
  chains: Chains,
  chainId: number
): chainId is Id {
  return chains.some((chain) => chain.id === chainId);
}

function getViemPublicClient(chainId: number, transport?: Transport) {
  const supportedChains = Object.values(chains);

  if (!isSupportedChainId(supportedChains, chainId)) {
    throw new Error("Chain not supported");
  }

  const chain = extractChain({ chains: supportedChains, id: chainId });
  if (!chain) {
    throw new Error("Chain config not found");
  }
  // default to public http endpoint
  transport ??= http();
  return createPublicClient({ chain, transport });
}
```

Reply:
> @pxrl You can opt out of viem's narrow type inference when some of your values are dynamic. Like here, your `chainId` is not strictly known at compile time so it is just a `number`.
> 
> Don't use viem's `extractChain` utility. It is just `Array.find` with static param checking
> 
> ```ts
> const chain: Chain | undefined = Object.values(chains).find((chain) => chain.id === chainId);
> ```

This works better and seems to build without problem. My editor (running tsserver in the background) opens the gates of hell on this one though. I can probably live with it.

![err](https://github.com/user-attachments/assets/dc097131-3857-462c-b13d-4dddae576438)